contracts/SponsorFaucet.sol

```diff
@@ -64,38 +109,180 @@ contract SponsorFaucet is Ownable, Pausable, ReentrancyGuard {
+        (bool success, ) = sponsor.call.value(amount)("");
+        require(success, "withdraw failed");
+    }
+
+    //set bounds for sponsorship
+    function setBounds(
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        gas_total_limit = gasTotalLimit;
+        collateral_total_limit = collateralTotalLimit;
+        gas_bound = gasBound;
+        collateral_bound = collateralBound;
+        upper_bound = upperBound;
+    }
+
+    //set bounds for general contracts
+    function setGeneralBounds(
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        general_gas_total_limit = gasTotalLimit;
+        general_collateral_total_limit = collateralTotalLimit;
+        general_gas_bound = gasBound;
+        general_collateral_bound = collateralBound;
+        general_upper_bound = upperBound;
+    }
+
+    //set bounds for special contracts
+    function setSpecialBounds(
+        address addr,
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        special_dapps[addr].gas_total_limit = gasTotalLimit;
+        special_dapps[addr].collateral_total_limit = collateralTotalLimit;
+        special_dapps[addr].gas_bound = gasBound;
+        special_dapps[addr].collateral_bound = collateralBound;
+        special_dapps[addr].upper_bound = upperBound;
+    }
+
+
+    /* ===== Public utility functions ===== */
+    function addGeneralContracts(address[] memory addAddr) public {
+        _onlyWhitelistAdmin();
+        for(uint256 i = 0; i < addAddr.length; i++) {
+            general_contracts.push(addAddr[i]);
+        }
+    }
+
+    function addSpecialContracts(address[] memory addAddr) public {
+        _onlyWhitelistAdmin();
+        for(uint256 i = 0; i < addAddr.length; i++) {
+            special_contracts.push(addAddr[i]);
+        }
+    }
+
+    function specialContractsCount() public view returns (uint256) {
+        return special_contracts.length;
+    }
+
+    function generalContractsCount() public view returns (uint256) {
+        return general_contracts.length;
+    }
+
+    function isGeneralContract(address dapp) public view returns (bool) {
+        for(uint256 i = 0; i < general_contracts.length; i++) {
```

Comment:
The worst case here needs general_contracts.length * gas_for_SLOAD gas in total. Will it exceed the gas limit of a transaction when general_contracts is too large?

Reply:
rm array and use mapping instead to avoid this.
rename: general => large, special => custom